vm/interpreter/src/circ_supply.rs

```diff
@@ -0,0 +1,227 @@
+
+    let six_months = 183 * network::EPOCHS_IN_DAY;
+    totals_by_epoch.insert(six_months, TokenAmount::from(82_717_041));
+
+    let one_year = 365 * network::EPOCHS_IN_DAY;
+    totals_by_epoch.insert(one_year, TokenAmount::from(22_421_712));
+
+    let two_years = 2 * 365 * network::EPOCHS_IN_DAY;
+    totals_by_epoch.insert(two_years, TokenAmount::from(7_223_364));
+
+    let three_years = 3 * 365 * network::EPOCHS_IN_DAY;
+    totals_by_epoch.insert(three_years, TokenAmount::from(87_637_883));
+
+    let six_years = 6 * 365 * network::EPOCHS_IN_DAY;
+    totals_by_epoch.insert(six_years, TokenAmount::from(400_000_000));
+
+    totals_by_epoch
+}
+
+fn init_genesis_info<DB: BlockStore>(bs: &DB) -> Result<GenesisInfo, String> {
+    let mut ignition = GenesisInfo::default();
+
+    let genesis_block = genesis(bs)
+        .map_err(|_| "Failed to get Genesis Block".to_string())?
+        .ok_or_else(|| "Genesis Block doesnt exist".to_string())?;
+
+    let gts =
+        Tipset::new(vec![genesis_block]).map_err(|_| "Failed to get Genesis Tipset".to_string())?;
+
+    // Parent state of genesis tipset is tipset state
+    let st = gts.parent_state();
+
+    let state_tree =
+        StateTree::new_from_root(bs, &st).map_err(|_| "Failed to load state tree".to_string())?;
+
+    ignition.genesis_market_funds = get_fil_market_locked(&state_tree)?;
+    ignition.genesis_pledge = get_fil_power_locked(&state_tree)?;
+
+    Ok(ignition)
+}
+
+pub fn setup_preignition_genesis_actors_testnet<DB: BlockStore>(
+    bs: &DB,
+) -> Result<GenesisInfo, String> {
+    let mut pre_ignition = init_genesis_info(bs)?;
+
+    let totals_by_epoch: HashMap<ChainEpoch, TokenAmount> = get_totals_by_epoch();
+
+    for (unlock_duration, initial_balance) in totals_by_epoch {
+        let ms = multisig::State {
+            signers: vec![],
+            num_approvals_threshold: 0,
+            next_tx_id: multisig::TxnID(0),
+            initial_balance,
+            start_epoch: ChainEpoch::default(),
+            unlock_duration,
+            pending_txs: Cid::default(),
+        };
+        pre_ignition.genesis_msigs.push(ms);
+    }
+
+    Ok(pre_ignition)
+}
+
+pub fn setup_postignition_genesis_actors_testnet<DB: BlockStore>(
+    bs: &DB,
+) -> Result<GenesisInfo, String> {
+    let mut post_ignition = init_genesis_info(bs)?;
+
+    let totals_by_epoch: HashMap<ChainEpoch, TokenAmount> = get_totals_by_epoch();
+
+    for (unlock_duration, initial_balance) in totals_by_epoch {
+        let ms = multisig::State {
+            signers: vec![],
+            num_approvals_threshold: 0,
+            next_tx_id: multisig::TxnID(0),
+            initial_balance: initial_balance * FILECOIN_PRECISION,
+            start_epoch: UPGRADE_LIFTOFF_HEIGHT,
+            unlock_duration,
+            pending_txs: Cid::default(),
+        };
+        post_ignition.genesis_msigs.push(ms);
+    }
+
+    Ok(post_ignition)
+}
```

Comment:
First, this adds a lot of complexity to handle the upgrade logic, and this isn't done correctly. FIrst, we don't have any use of GenesisInfo, but also you're porting over go specific things that will cause issues for us.

Reply:
Kinda confused by what you mean here. As I have it right now the GenesisInfo is used in get_fil_vested